File: app/src/main/java/com/frankegan/verdant/FABToggle.java
package com.frankegan.verdant;

import android.content.Context;
import android.support.v4.view.ViewCompat;
import android.util.AttributeSet;
import android.widget.Checkable;
import android.widget.ImageButton;

/**
 * Created by frankegan on 5/7/16.
 *
 * A {@link Checkable} {@link ImageButton} which has a minimum offset i.e. translation Y.
 */
public class FABToggle extends ImageButton implements Checkable {

    private static final int[] CHECKED_STATE_SET = { android.R.attr.state_checked };

    private boolean isChecked = false;
    private int minOffset;

    public FABToggle(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setOffset(int offset) {
        offset = Math.max(minOffset, offset);
        if (getTranslationY() != offset) {
            setTranslationY(offset);
            ViewCompat.postInvalidateOnAnimation(this);
        }
    }

    public void setMinOffset(int minOffset) {
        this.minOffset = minOffset;
    }

    @Override
    public boolean isChecked() {
        return isChecked;
    }

    @Override
    public void setChecked(boolean isChecked) {
        if (this.isChecked != isChecked) {
            this.isChecked = isChecked;
            refreshDrawableState();
        }
    }

    @Override
    public void toggle() {
        setChecked(!isChecked);
    }

    @Override
    public int[] onCreateDrawableState(int extraSpace) {
        final int[] drawableState = super.onCreateDrawableState(extraSpace + 1);
        if (isChecked()) {
            mergeDrawableStates(drawableState, CHECKED_STATE_SET);
        }
        return drawableState;
    }

}

File: app/src/main/java/com/frankegan/verdant/FABToggle.kt
package com.frankegan.verdant

import android.content.Context
import android.support.v7.widget.AppCompatImageButton
import android.util.AttributeSet
import android.view.View
import android.widget.Checkable
import android.widget.ImageButton

/**
 * Created by frankegan on 5/7/16.
 *
 * A [Checkable] [ImageButton] which has a minimum offset i.e. translation Y.
 */
class FABToggle(context: Context, attrs: AttributeSet) : AppCompatImageButton(context, attrs), Checkable {

    private var isChecked = false

    override fun isChecked(): Boolean {
        return isChecked
    }

    override fun setChecked(isChecked: Boolean) {
        if (this.isChecked != isChecked) {
            this.isChecked = isChecked
            refreshDrawableState()
        }
    }

    override fun toggle() {
        setChecked(!isChecked)
    }

    override fun onCreateDrawableState(extraSpace: Int): IntArray {
        val drawableState = super.onCreateDrawableState(extraSpace + 1)
        if (isChecked()) {
            View.mergeDrawableStates(drawableState, CHECKED_STATE_SET)
        }
        return drawableState
    }

    companion object {
        private val CHECKED_STATE_SET = intArrayOf(android.R.attr.state_checked)
    }

}